src/Language/PureScript/Sugar/TypeClasses.hs

```diff
@@ -323,11 +333,12 @@ typeInstanceDictionaryDeclaration sa@(ss, _) name mn deps className tys decls =
       -- Create the type of the dictionary
       -- The type is a record type, but depending on type instance dependencies, may be constrained.
       -- The dictionary itself is a record literal.
-      let superclasses = superClassDictionaryNames typeClassSuperclasses `zip`
-            [ Abs (VarBinder ss UnusedIdent) (DeferredDictionary superclass tyArgs)
-            | (Constraint _ superclass _ suTyArgs _) <- typeClassSuperclasses
-            , let tyArgs = map (replaceAllTypeVars (zip (map fst typeClassArguments) tys)) suTyArgs
-            ]
+      tys' <- traverse (replaceAllTypeSynonymsM syns kinds) tys
+      superclassesDicts <- for typeClassSuperclasses $ \(Constraint _ superclass _ suTyArgs _) -> do
+        suTyArgs' <- traverse (replaceAllTypeSynonymsM syns kinds) suTyArgs
```

Comment:
Unless I'm missing something, this `kinds` that got passed all over the place is always [`mempty`](https://github.com/purescript/purescript/pull/3965/files#diff-b2c45abe2b96a6ae2e0da4d769b691e5bc4ee0e1d86e46d9b973df822d10a698R81). I'm not sure what the consequences of that are, but if this is the intended behavior, why not just use `mempty` here?

Reply:
I kept a `kinds` binding while applying your other suggestion but attached to it a comment rephrasing what Nate said in https://github.com/purescript/purescript/pull/3779 about the changes to the `Sugar.TypeClasses` module:

> Since this happens before type-checking (and thus before kind-checking), there's some dodgy code to expand type synonyms before kinds are known. I don't know if this is a problem in practice (you might just get garbage ill-kinded results and bad errors). My opinion is that this desugaring happens too soon in the pipeline and should happen after other type declarations are kind checked.

Reply:
That's good enough for me!